### PR TITLE
[cosmwasm] instantiate contracts on multiple chains

### PR DESCRIPTION
**Pull request by 0xfirefist**
This PR contains:
- code that adds stages to the wormhole pipeline. Stages to upgrade wormhole guardian set.
- script for instantiating pyth
- deployment results for chains

### CODE REVIEW

File: target_chains/cosmwasm/tools/src/helper.ts [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+  // clear tmp directory before downloading contracts
+  rimrafSync(tmpCodeStorageDir);
+
+  if (existsSync(tmpCodeStorageDir) === false) {
+    mkdirSync(tmpCodeStorageDir, { recursive: true });
+  }
+
+  const uniqueArtifactsZipName = Array.from(new Set(artifactZipFileNames));
+
+  // download zip files
```
Comment (ali-bahjati) [inline target_chains/cosmwasm/tools/src/helper.ts:27]: I might be a bit paranoid, but can we also verify the checksum? If one of our gh accounts get compromised they can swap the binary without us knowing it.

File: target_chains/cosmwasm/tools/src/wormhole-stub.ts [hunk 2]
```diff
@@ -123,19 +108,36 @@ async function run() {
       },
     });
 
+    // 4 vaa upgrades for guardian set
+    Object.keys(VAA_UPGRADES).forEach((id) => {
```
Comment (ali-bahjati) [inline target_chains/cosmwasm/tools/src/wormhole-stub.ts:112]: For mainnet only right?

File: target_chains/cosmwasm/tools/src/chains-manager/chains.ts [hunk 3]
```diff
@@ -96,16 +97,23 @@ export const ChainsConfigTestnet: Record<ChainIdTestnet, ChainConfig> = {
...
 };
 
 /**
- * This method will return an executor for that corresponding chainType for given chainId.
+ * This method will return an executor for given chainConfig.
  */
-export function createExecutorForChain(
-  chainId: ChainIdTestnet,
+export function createExecutorForChain<T extends ChainConfig>(
+  chainConfig: T,
```
Comment (jayantk) [inline target_chains/cosmwasm/tools/src/chains-manager/chains.ts:114]: shouldn't this be `chainConfig: ChainConfig`? You don't use the `T` beyond that

File: target_chains/cosmwasm/tools/src/helper.ts (hunk 1 above)
Comment (jayantk) [reply target_chains/cosmwasm/tools/src/helper.ts:27]: yeah agree with this

File: target_chains/cosmwasm/tools/src/helper.ts [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+// We have released the compile contacts on github. If a chain needs some specific
+// feature in a contract, a version of the contract with that specific features is
+// also released. For example, "injective.zip" for injective.
+// Given a contract version (eg 1.2.0) and  zipFileNames (eg ["injective", "osmosis"])
+// this method will return an object with key = zipFileName and value = compiledCode
+export async function getContractBytesDict(
+  artifactZipFileNames: string[],
+  version: string
+) {
+  const githubArtifactsLink = `https://github.com/pyth-network/pyth-crosschain/releases/download/pyth-cosmwasm-contract-v${version}/`;
```
Comment (jayantk) [inline target_chains/cosmwasm/tools/src/helper.ts:16]: i think it's a little questionable to hardcode this URL here. This ties the implementation to our git repository url, which is weird. It seems like this should be configurable and passed in. I don't feel strongly about this comment in this instance though -- no one else is going to run this -- so feel free to ignore.

I think it's also worthwhile clarifying exactly what expectations this function has on the zip files (e.g., it needs the wasm artifact inside to have a specific name, etc)

File: target_chains/cosmwasm/tools/src/helper.ts [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+// This method returns the file name which stores the deployment result for
+// wormhole stub
+export function getWormholeFileName(
+  chainId: string,
+  version: string,
+  mainnet: boolean
+): string {
+  const WORMHOLE_STORAGE_DIR = mainnet
+    ? "../wormhole-stub/mainnet"
+    : "../wormhole-stub/testnet";
```
Comment (jayantk) [inline target_chains/cosmwasm/tools/src/helper.ts:85]: extract constant variables to hold these string constants. that way you can rename the directories in code without chasing down all of the instances of the string "wormhole-stub"

Comment (jayantk) [reply target_chains/cosmwasm/tools/src/helper.ts:85]: also why is this filepath "wormhole-stub/testnet" and the pyth one is "testnet/instantiate-pyth"? seems like those should be in the same order

File: target_chains/cosmwasm/tools/src/instantiate-pyth.ts [hunk 6]
```diff
@@ -0,0 +1,164 @@
...
+    governance_source_index: 0,
+    governance_sequence_number: 0,
+    chain_id: wormholeChainId,
+    valid_time_period_secs: 60,
+    fee: {
+      amount: "1",
+      denom: feeDenom,
+    },
+    data_sources: [
+      {
```
Comment (jayantk) [inline target_chains/cosmwasm/tools/src/instantiate-pyth.ts:139]: maybe comment here that all chains (including testnets) are using the mainnet data sources

File: target_chains/cosmwasm/tools/src/wormhole-stub.ts (hunk 2 above)
Comment (jayantk) [reply target_chains/cosmwasm/tools/src/wormhole-stub.ts:112]: dev is doing the "testnets read mainnet feeds" thing we discussed for cosmwasm

File: target_chains/cosmwasm/tools/src/helper.ts (hunk 5 above)
Comment (jayantk) [reply target_chains/cosmwasm/tools/src/helper.ts:85]: also, might be nice to write down what filepaths this generates and what gets stored there. Could be a good code comment on wherever you put the string constants.

**Review by jayantk — COMMENTED** [review]
nice. I have mostly minor comments. The one thing I would like to see is the checksum checking, because there is a potential security hole there (and we should be paranoid about this stuff). Let's write down the expected checksums in this repo as a config file and then we can check the downloads against that.

File: target_chains/cosmwasm/tools/src/wormhole-stub.ts (hunk 2 above)
Comment (0xfirefist) [reply target_chains/cosmwasm/tools/src/wormhole-stub.ts:112]: testnet will read from mainnet feeds

File: target_chains/cosmwasm/tools/src/helper.ts (hunk 5 above)
Comment (0xfirefist) [reply target_chains/cosmwasm/tools/src/helper.ts:85]: Extracting constant variables. This is the only place it is being used and hence the only place we would need to make change.

File: target_chains/cosmwasm/tools/src/helper.ts (hunk 1 above)
Comment (0xfirefist) [reply target_chains/cosmwasm/tools/src/helper.ts:27]: hardcoded checksum

File: target_chains/cosmwasm/tools/src/helper.ts (hunk 4 above)
Comment (0xfirefist) [reply target_chains/cosmwasm/tools/src/helper.ts:18]: Clarified the comment.

**Review by ali-bahjati — APPROVED** [review]

**Review by ali-bahjati — APPROVED** [review]

**Pull request merged**